src/libsyntax/parse/parser.rs

```diff
@@ -6214,14 +6214,24 @@ impl<'a> Parser<'a> {
         })
     }
 
-    pub fn ensure_complete_parse<F>(&mut self, allow_semi: bool, on_err: F)
-        where F: FnOnce(&Parser)
+    pub fn ensure_complete_parse(&mut self, kind: ExpansionKind, macro_name: ast::Name, span: Span)
     {
-        if allow_semi && self.token == token::Semi {
+        // We allow semicolons at the end of expressions -- e.g. the semicolon in
+        // `macro_rules! m { () => { panic!(); } }` isn't parsed by `.parse_expr()`,
+        // but `m!()` is allowed in expression positions (c.f. issue #34706).
```

Comment:
Also, ignoring the trailing semicolon creates unexpected (to me) behavior today -- for example, in [this gist](https://play.rust-lang.org/?gist=f71bc58be56c08b75a471fdb7ab6b696&version=nightly&backtrace=0) I would expect the expansion to have the unit type instead of `i32`.